client/question/services/FeedbackGeneratorService.js

```diff
@@ -24,12 +24,60 @@ tie.factory('FeedbackGeneratorService', [
   'RUNTIME_ERROR_FEEDBACK_MESSAGES', 'WRONG_LANGUAGE_ERRORS', 'LANGUAGE_PYTHON',
   'CLASS_NAME_AUXILIARY_CODE', 'CLASS_NAME_SYSTEM_CODE', 'PARAGRAPH_TYPE_TEXT',
   'PARAGRAPH_TYPE_CODE', 'PARAGRAPH_TYPE_SYNTAX_ERROR',
+  'PYTHON_PRIMER_BUTTON_NAME', 'EXCLUDE_LANGUAGE_UNFAMILARITY_ERROR',
+  'EXCLUDE_CONSECUTIVE_SAME_RUNTIME_ERROR', 'UNFAMILIARITY_THRESHOLD',
   function(
     FeedbackObjectFactory, TranscriptService, ReinforcementGeneratorService,
     CODE_EXECUTION_TIMEOUT_SECONDS, SUPPORTED_PYTHON_LIBS,
     RUNTIME_ERROR_FEEDBACK_MESSAGES, WRONG_LANGUAGE_ERRORS, LANGUAGE_PYTHON,
     CLASS_NAME_AUXILIARY_CODE, CLASS_NAME_SYSTEM_CODE, PARAGRAPH_TYPE_TEXT,
-    PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR) {
+    PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR,
+    PYTHON_PRIMER_BUTTON_NAME, EXCLUDE_LANGUAGE_UNFAMILARITY_ERROR,
+    EXCLUDE_CONSECUTIVE_SAME_RUNTIME_ERROR, UNFAMILIARITY_THRESHOLD) {
+
+    /**
+     * Counter to keep track of language unfamiliarity errors which include
+     * syntax errors and wrong language errors.
+     *
+     * @type {number}
+     */
+    var consecutiveLanguageUnfamiliarityCounter = 0;
+
+    /**
+     * Counter to keep track of consecutive same errors.
+     *
+     * @type {number}
+     */
+    var consecutiveSameRuntimeErrorCounter = 0;
+
+    /**
+     * Variable to store the error string immediately before the current error.
+     * Will be used to see if the user is receiving the same exact error
+     * consecutively, a possible indication of language unfamiliarity.
+     *
+     * @type {string}
+     */
+    var previousErrorString = '';
+
+    /**
+     * Resets all but one specific counter or all counters, depending on
+     * the parameter.
+     *
+     * @param {string} currentCounter
+     * @private
+     */
+    var _resetCounters = function(currentCounter) {
+      // If the parameter is null, reset all counters.
+      var counterNotToReset = currentCounter || '';
+      if (counterNotToReset !== EXCLUDE_LANGUAGE_UNFAMILARITY_ERROR) {
+        consecutiveLanguageUnfamiliarityCounter = 0;
+      }
+      if (counterNotToReset !== EXCLUDE_CONSECUTIVE_SAME_RUNTIME_ERROR) {
+        consecutiveSameRuntimeErrorCounter = 0;
+        previousErrorString = '';
```

Comment:
Should this be previousRuntimeErrorString (since you're dealing with a specific class of error here)?

That said, this probably shouldn't even be here -- it's outside what this function's supposed to do. Could the caller do it instead?